README.md
# ANSi View (PS3)

[![Downloads][img_downloads]][app_downloads] [![Release][img_latest]][app_latest] [![License][img_license]][app_license]

**ANSi View** is an application to view ANSi and artscene related file formats on the PlayStation 3.

The `ansiview-ps3` homebrew app supports viewing ANSI and ASCII files directly on your PS3.

![image](./docs/screenshots/screenshot_main.png)

**Comments, ideas, suggestions?** You can contact [me](https://github.com/bucanero/) on [Twitter](https://twitter.com/dparrino) and on [my website](http://www.bucanero.com.ar/).

# Features

The following formats are supported:

 - `.ANS` - ANSi (ANSI escape sequences: ANSI X3.64 standard)
 - `.PCB` - PCBoard Bulletin Board System (BBS) own file format
 - `.BIN` - Binary format (raw memory copy of text mode video memory)
 - `.ADF` - Artworx format, supporting custom character sets and palettes
 - `.IDF` - iCE Draw format, supporting custom character sets and palettes
 - `.TND` - TundraDraw format, supporting 24-bit color mode
 - `.XB` - The eXtended Binary XBin format, supporting custom character sets and palettes
 - `.TXT`, `.ASC`, `.NFO`, `.DIZ` - ASCII text files
 
 Supported Background music formats:
 
 - `.669` - Composer 669, Unis 669
 - `.AMF` - DSMI Advanced Module Format
 - `.DSM` - DSIK internal format
 - `.FAR` - Farandole Composer
 - `.GDM` - General DigiMusic
 - `.IT` - Impulse Tracker
 - `.IMF` - Imago Orpheus
 - `.MOD` - 15/31 instruments
 - `.MED` - OctaMED
 - `.MTM` - MultiTracker Module editor
 - `.OKT` - Amiga Oktalyzer
 - `.S3M` - Scream Tracker 3
 - `.STM` - Scream Tracker
 - `.STX` - Scream Tracker Music Interface Kit
 - `.ULT` - UltraTracker
 - `.APUN` - APlayer and UNI (MikMod)
 - `.XM` - FastTracker 2

# Download

Get the [latest version here][app_latest].

### Changelog

See the [latest changes here](CHANGELOG.md).

# Setup instructions

No special setup is needed. Just download the latest `ansiview-ps3.pkg` package and install it on your PlayStation 3.

# Usage

Using the application is simple and straight-forward: 

 - Move **UP/DOWN** to select the ANSi art file you want to view, and press ![X button](https://github.com/bucanero/pkgi-ps3/raw/master/data/CROSS.png). The selected image will scroll on the screen.
 - To see the item's SAUCE details, press ![Triangle](https://github.com/bucanero/pkgi-ps3/raw/master/data/TRIANGLE.png).
It will open the context menu on the screen. Press ![O button](https://github.com/bucanero/pkgi-ps3/raw/master/data/CIRCLE.png) to return to the list.
 - To reload the folder list, press ![Square](https://github.com/bucanero/pkgi-ps3/raw/master/data/SQUARE.png).
 - Press left **(L1/L2)** or right trigger buttons **(R1/R2)** to move pages up or down.

## Notes

- **USB art packs:** your folders must be placed on `/dev_usb000/ANSIVIEW/` or `/dev_usb001/ANSIVIEW/`.
- **HDD art packs:** folders will be scanned from `/dev_hdd0/game/NP0ANSIVW/USRDIR/DATA/`.

# 16colors ANSI/ASCII archive

The application provides direct access to the [16colors archive](https://16colo.rs) of artscene files. You can browse their entire art collection (1990 - 2020) if your PS3 has a network connection available.

**Note:** Downloaded art files are cached on `/dev_hdd0/game/NP0ANSIVW/USRDIR/CACHE/`.

# Credits

* [Bucanero](http://www.bucanero.com.ar/): Project developer
* [Dnawrkshp](https://github.com/Dnawrkshp/): [Artemis PS3](https://github.com/Dnawrkshp/ArtemisPS3)
* [ByteProject](https://github.com/ByteProject)/[bricas](https://github.com/bricas)/[fcambus](https://github.com/fcambus): [libansilove](https://github.com/ansilove/libansilove)

# Building

You need to have installed:

- [PS3 toolchain](https://github.com/bucanero/ps3toolchain)
- [PSL1GHT](https://github.com/bucanero/PSL1GHT) SDK
- [Tiny3D](https://github.com/Estwald/PSDK3v2/tree/master/libraries-src/Tiny3D) library (from Estwald)
- [MikMod](https://github.com/ps3dev/ps3libraries/blob/master/scripts/011-libmikmod-3.1.11.sh) library
- [libansilove](https://github.com/bucanero/libansilove) library (my own custom version)
- [libcurl](https://github.com/ps3dev/ps3libraries/blob/master/scripts/016-libcurl-7.31.0.sh) library
- [dbglogger lib](https://github.com/bucanero/psl1ght-libs/tree/master/dbglogger) (my own debug logging library)

Run `make` to create a release build. After that, run `make pkg` to create a `.pkg` install file. 

You can also set the `PS3LOAD` environment variable to the PS3 IP address: `export PS3LOAD=tcp:x.x.x.x`.
This will allow you to use `make run` and send `ansiview-ps3.self` directly to the PS3Load listener.

To enable debug logging, build ANSi View with `make DEBUGLOG=1`. The application will send debug messages to
UDP multicast address `239.255.0.100:30000`. To receive them you can use [socat][] on your computer:

    $ socat udp4-recv:30000,ip-add-membership=239.255.0.100:0.0.0.0 -

# License

`ansiview-ps3` is released under the [MIT License](LICENSE).

[PSDLE]: https://repod.github.io/psdle/
[socat]: http://www.dest-unreach.org/socat/
[app_downloads]: https://github.com/bucanero/ansiview-ps3/releases
[app_latest]: https://github.com/bucanero/ansiview-ps3/releases/latest
[app_license]: https://github.com/bucanero/ansiview-ps3/blob/master/LICENSE
[img_downloads]: https://img.shields.io/github/downloads/bucanero/ansiview-ps3/total.svg?maxAge=3600
[img_latest]: https://img.shields.io/github/release/bucanero/ansiview-ps3.svg?maxAge=3600
[img_license]: https://img.shields.io/github/license/bucanero/ansiview-ps3.svg?maxAge=2592000


CHANGELOG.md
# Changelog

All notable changes to the `ansiview-ps3` project will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]()

## [v0.5.0](https://github.com/bucanero/ansiview-ps3/releases/tag/v0.5.0) - 2020-05-15

First public release.


include/files.h
#ifdef DEBUG_ENABLE_LOGGING
#include <dbglogger.h>
#define LOG dbglogger_log
#else
#define LOG(...)
#define dbglogger_init(...)
#endif

#include "sauce.h"
#include "list.h"

#define ANSIVIEW_PATH			"/dev_hdd0/game/NP0ANSIVW/USRDIR/"
#define ANSIVIEW_DATA_PATH		ANSIVIEW_PATH "DATA/"
#define ANSIVIEW_MUSIC_PATH		ANSIVIEW_PATH "MUSIC/"
#define ANSIVIEW_LOCAL_CACHE	ANSIVIEW_PATH "CACHE/"
#define ANSIVIEW_UPDATE_URL		"https://api.github.com/repos/bucanero/ansiview-ps3/releases/latest"

#define USB0_PATH               "/dev_usb000/"
#define USB1_PATH               "/dev_usb001/"
#define ANSIVIEW_PATH_USB0		USB0_PATH "ANSIVIEW/"
#define ANSIVIEW_PATH_USB1		USB1_PATH "ANSIVIEW/"

#define EXPORT_PATH_USB0        USB0_PATH "EXPORT/"
#define EXPORT_PATH_USB1        USB1_PATH "EXPORT/"

#define ONLINE_URL				"ftp://16colo.rs/pack/"
#define ONLINE_CACHE_TIMEOUT    24*3600     // 1-day local cache

// Save flags
#define SAVE_FLAG_LOCKED        1
#define FOLDER_FLAG_FTP         2
#define FILE_FLAG_SAUCE         4
#define SAVE_FLAG_PS3           8
#define FOLDER_FLAG_YEAR        16
#define FILE_FLAG_PSP           32
#define FILE_FLAG_PSV           64

enum char_flag_enum
{
    CHAR_TAG_NULL,
    CHAR_TAG_PS1,
    CHAR_TAG_PS2,
    CHAR_TAG_PS3,
    CHAR_TAG_PSP,
    CHAR_TAG_PSV,
    CHAR_TAG_APPLY,
    CHAR_TAG_OWNER,
    CHAR_TAG_LOCKED,
    CHAR_RES_TAB,
    CHAR_RES_LF,
    CHAR_TAG_TRANSFER,
    CHAR_TAG_ZIP,
    CHAR_RES_CR,
    CHAR_TAG_PCE,
    CHAR_TAG_WARNING,
    CHAR_BTN_X,
    CHAR_BTN_S,
    CHAR_BTN_T,
    CHAR_BTN_O,
};

typedef struct option_entry
{
    char * line;
    char * * name;
    char * * value;
    int id;
    int size;
    int sel;
} option_entry_t;

typedef struct file_entry
{
    unsigned char type;
    char * name;
    char * file;
    unsigned char activated;
//	unsigned int flags;
    sauce_record_t * sauce;
    char * details;
    int options_count;
    option_entry_t * options;
} file_entry_t;

typedef struct folder_entry
{
    char * name;
	char * title_id;
	char * path;
	unsigned int flags;
    int file_count;
    file_entry_t * files;
} folder_entry_t;

typedef struct
{
    folder_entry_t * folders;
    int count;
    char path[128];
    char* title;
    unsigned char icon_id;
    void (*UpdatePath)(char *);
    int (*ReadFiles)(folder_entry_t *);
    folder_entry_t* (*ReadFolders)(const char*, int *);
} menu_list_t;

folder_entry_t * ReadLocalFolderList(const char* userPath, int * gmc);
folder_entry_t * ReadOnlineList(const char* urlPath, int * gmc);
void UnloadFolderList(folder_entry_t * list, int count);
int isGameActivated(folder_entry_t game);
char * ParseActivatedGameList(folder_entry_t * list, int count);
char * readFile(const char * path, long* size);
int qsortFolderList_Compare(const void* A, const void* B);
int qsortFileList_Compare(const void* A, const void* B);
long getFileSize(const char * path);
int ReadFiles(folder_entry_t * folder);
int ReadOnlineFiles(folder_entry_t * game);
int LoadMusicList(list_t* m_list);
int ReadMusicFiles(folder_entry_t * folder, list_t* m_list);

int http_init(void);
void http_end(void);
int http_download(const char* url, const char* filename, const char* local_dst, int show_progress);
int ftp_util(const char *ftp_url, const char* dst_name);

int extract_zip(const char* zip_file, const char* dest_path);
int zip_directory(const char* basedir, const char* inputdir, const char* output_zipfile);

int show_dialog(int dialog_type, const char * str);
void init_progress_bar(const char* progress_bar_title, const char* msg);
void update_progress_bar(long unsigned int* progress, const long unsigned int total_size, const char* msg);
void end_progress_bar(void);

int init_loading_screen(const char* msg);
void stop_loading_screen();

int apply_cheat_patch_code(const char* fpath, const char* title_id, file_entry_t* code);
